hash/hash.go

```diff
@@ -8,51 +8,32 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
-	"hash"
 
-	"github.com/mesg-foundation/engine/hash/structhash"
+	"github.com/mesg-foundation/engine/hash/hashserializer"
 	"github.com/mr-tron/base58"
 )
 
-// DefaultHash is a default hashing algorithm - "sha256".
-var DefaultHash = sha256.New
+// size is the default size for the hashing algorithm.
+const size = sha256.Size
 
-// size is a default size for hashing algorithm.
-var size = DefaultHash().Size()
+// sumFunc is the default function to hash.
+var sumFunc = sha256.Sum256
 
 var errInvalidLen = errors.New("hash: invalid length")
 
-// Digest represents the partial evaluation of a checksum.
-type Digest struct {
-	hash.Hash
-}
-
-// Sum appends the current checksum to b and returns the Hash.
-func (d *Digest) Sum(b []byte) Hash {
-	return Hash(d.Hash.Sum(b))
-}
-
-// WriteObject  adds an interface data to the running hash.
-// It never retruns an error.
-func (d *Digest) WriteObject(v interface{}) (int, error) {
-	return d.Write(structhash.Dump(v))
-}
-
 // A Hash is a type for representing common hash.
 type Hash []byte
 
-// New returns new hash from a given integer.
-func New() *Digest {
-	return &Digest{
-		Hash: DefaultHash(),
-	}
+// Dump takes a structure that implement HashSerializable and returns its hash.
+func Dump(v hashserializer.HashSerializable) Hash {
+	h := sumFunc([]byte(v.HashSerialize()))
+	return Hash(h[:])
 }
 
-// Dump takes an interface and returns its hash representation.
-func Dump(v interface{}) Hash {
-	d := New()
-	d.WriteObject(v)
-	return d.Sum(nil)
+// Sum takes a slice of byte and returns its hash.
+func Sum(v []byte) Hash {
```

Comment:
how can `bytes.Equal` replace a sha256 sum function?

Reply:
as I can see it is used like
```go
require.True(t, hash.Sum(structValueSort1).Equal(hash.Sum(structValueSort2))) 
```

And this could be written as 
```go
require.True(t, bytes.Equals(structValueSort1, structValueSort2))
```

Because you don't need to convert it to hash to compare the marshal output.

Reply:
the struct are not equals, but their hash-serialized is

Reply:
But here you can pass and compare hash for two byte slices. We can keep it if you feel there a need for that. Just leaving the change and tests below

```
➜  types git:(feature/simple-hash) ✗ git diff
diff --git a/protobuf/types/struct_test.go b/protobuf/types/struct_test.go
index b60ed9d0..c2ed0177 100644
--- a/protobuf/types/struct_test.go
+++ b/protobuf/types/struct_test.go
@@ -1,6 +1,7 @@
 package types
 
 import (
+       bytes "bytes"
        "testing"
 
        "github.com/cosmos/cosmos-sdk/codec"
@@ -91,7 +92,7 @@ func TestStructMarshal(t *testing.T) {
                        require.NoError(t, err)
                        structValueSort2, err = cdc.MarshalBinaryLengthPrefixed(structSort2)
                        require.NoError(t, err)
-                       require.True(t, hash.Sum(structValueSort1).Equal(hash.Sum(structValueSort2)))
+                       require.True(t, bytes.Equal(structValueSort1, structValueSort2))
                })
                t.Run("Unmarshal", func(t *testing.T) {
                        require.NoError(t, cdc.UnmarshalBinaryLengthPrefixed(structValueSort1, &structUnm1))
@@ -114,7 +115,7 @@ func TestStructMarshal(t *testing.T) {
                        require.NoError(t, err)
                        structValueSort2, err = cdc.MarshalJSON(structSort2)
                        require.NoError(t, err)
-                       require.True(t, hash.Sum(structValueSort1).Equal(hash.Sum(structValueSort2)))
+                       require.True(t, bytes.Equal(structValueSort1, structValueSort2))
                })
                t.Run("Unmarshal", func(t *testing.T) {
                        require.NoError(t, cdc.UnmarshalJSON(structValueSort1, &structUnm1))
➜  types git:(feature/simple-hash) ✗ go test
PASS
ok  	github.com/mesg-foundation/engine/protobuf/types	0.042s
```

Reply:
I changed the test but I would like to keep the function as it is useful in an upcoming PR ;)